### PR TITLE
[GCE Discovery] Automatically set project-id and zone

### LINKED ISSUE #13618: Automatically set project-id and zone
The project id can be retrieved from the running instance by calling  `http://metadata.google.internal/computeMetadata/v1/project/project-id`.

Same for the default zone: `http://metadata.google.internal/computeMetadata/v1/project/attributes/google-compute-default-zone
`

No need then to force them in `elasticsearch.yml`.

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**
Settings `cloud.gce.project_id`, `cloud.gce.zone` are not mandatory - they could be automatically set from a gce metadata server.  Picking up the default zone is useful for cases like a single region. 

Details on impl differences with GCE:
[com.google.cloud.ServiceOptions#getDefaultProjectId ](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java#L333) tries to resolve project based on
* system property `GOOGLE_CLOUD_PROJECT` (or legacy system property `GCLOUD_PROJECT`) 
* environment variable `GOOGLE_CLOUD_PROJECT` or `GCLOUD_PROJECT` 
* from `http://metadata.google.internal/computeMetadata/v1/project/project-id`

This PR allows without settings to fallback to retrieve info from gce metadata server  

Closes #13618

### CODE REVIEW

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by bleskes** [comment]
@vladimirdolzhenko thanks for picking this up! do you mind updating the PR title to indicate it relates to GCE discovery and maybe add a sentence or two to the descriptiont telling people what zone and project id are used for and why this is a good default?

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java [hunk 1]
```diff
@@ -103,9 +108,58 @@
...
+            final String defaultZone =
+                getAppEngineValueFromMetadataServer("/computeMetadata/v1/project/attributes/google-compute-default-zone");
+            return Collections.singletonList(defaultZone);
+        } catch (Exception e) {
+            logger.warn("unable to resolve default zone from metadata server for GCE discovery service", e);
+        }
+        return null;
+    }
+
+    private String getAppEngineValueFromMetadataServer(String serviceURL) throws GeneralSecurityException, IOException {
```
Comment (tlrx) [inline plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: Can we add a bit of doc and why we are copying the SDK code here?

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java [hunk 2]
```diff
@@ -103,9 +108,58 @@
...
+        HttpRequestFactory requestFactory = httpTransport.createRequestFactory();
+        HttpRequest request = requestFactory.buildGetRequest(url)
+            .setConnectTimeout(500)
+            .setReadTimeout(500)
+            .setHeaders(new HttpHeaders().set("Metadata-Flavor", "Google"));
+        HttpResponse response = Access.doPrivilegedIOException(() -> request.execute());
+        return headerContainsMetadataFlavor(response) ? response.parseAsString() : null;
+    }
+
+    static boolean headerContainsMetadataFlavor(HttpResponse response) {
```
Comment (tlrx) [inline plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: Can we add a simple test for this?

File: docs/plugins/discovery-gce.asciidoc [hunk 3]
```diff
@@ -26,12 +26,14 @@ The following gce settings (prefixed with `cloud.gce`) are supported:
 
  `project_id`::
 
-     Your Google project id (mandatory).
+     Your Google project id.
+     If the value is not specified, default project id is picked up from a metadata server.
```
Comment (tlrx) [inline docs/plugins/discovery-gce.asciidoc]: Can we add a note that the GOOGLE env var are not picked up?

**Review by tlrx — COMMENTED** [review]
I left some minor comments. As a follow up I think we should upgrade the SDK here. And do you think that we could have a fixture based integration test (QA) for this plugin?

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java (hunk 1 above)
Comment (vladimirdolzhenko) [reply plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: ++

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java (hunk 2 above)
Comment (vladimirdolzhenko) [reply plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: ++

File: docs/plugins/discovery-gce.asciidoc (hunk 3 above)
Comment (vladimirdolzhenko) [reply docs/plugins/discovery-gce.asciidoc]: ++

**Review by tlrx — APPROVED** [review]
LGTM - Did you test it on GCE?

**Comment by dadoonet** [comment]
May be that change since I opened the original issue. :)

**Comment by vladimirdolzhenko** [comment]
`google-compute-default-zone` metaattribute appeared after specifying default zone for the project :
screen shot 2018-09-20 at 22 06 27

`AccessControlException` happens on testing in a real GCE, to fix it QA test was added (two settings - `cloud.gce.host` and `cloud.gce.root_url`  are exposed for that).

**Comment by vladimirdolzhenko** [comment]
@elasticmachine retest it please

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@elasticmachine test this please

**Comment by vladimirdolzhenko** [comment]
@tlrx I'd like to ask you to have another look into this PR as I've spotted an issue on a real GCE and therefore added QA tests (and performed manual tests after fix). 

One think I don't feel quite comfortable is to exposing two new settings - it is not possible to run QA cluster w/o them.

File: plugins/discovery-gce/qa/gce/build.gradle [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+apply plugin: 'elasticsearch.rest-test'
+
+final int gceNumberOfNodes = 3
+File gceDiscoveryFile =  new File(project.buildDir, 'generated-resources/nodes.uri')
+
+dependencies {
+    testCompile project(path: ':plugins:discovery-gce', configuration: 'runtime')
+}
+
+/** A task to start the GCEFixture which emulates an GCE service **/
```
Comment (tlrx) [inline plugins/discovery-gce/qa/gce/build.gradle]: nit: an -> a

File: plugins/discovery-gce/qa/gce/src/test/java/org/elasticsearch/cloud/gce/GCEFixture.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/**
+ * {@link GCEFixture} is a fixture that emulates an GCE service.
```
Comment (tlrx) [inline plugins/discovery-gce/qa/gce/src/test/java/org/elasticsearch/cloud/gce/GCEFixture.java]: Same here

File: plugins/discovery-gce/qa/gce/src/test/java/org/elasticsearch/cloud/gce/GCEFixture.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                + TOKEN + "\", \"expires_in\":3268, \"token_type\":\"" + TOKEN_TYPE + "\"}"));
+
+        // https://cloud.google.com/compute/docs/reference/rest/v1/instances
+        handlers.insert(authPath(HttpGet.METHOD_NAME, "/compute/v1/projects/{project}/zones/{zone}/instances"),
+            request -> {
+                final Map<String, String> headers = new HashMap<>(JSON_CONTENT_TYPE);
+                commonHeaderConsumer.accept(headers);
+
+                final StringBuilder builder = new StringBuilder();
+                builder.append("{\"id\": \"test-instances\",\"items\":[");
```
Comment (tlrx) [inline plugins/discovery-gce/qa/gce/src/test/java/org/elasticsearch/cloud/gce/GCEFixture.java]: Could you have used a jsonBuilder() instead?

File: plugins/discovery-gce/src/test/java/org/elasticsearch/cloud/gce/GceInstancesServiceTests.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+import static org.hamcrest.Matchers.is;
+
+@SuppressForbidden(reason = "use http server")
+public class GceInstancesServiceTests extends ESTestCase {
+    private static HttpServer httpServer;
```
Comment (tlrx) [inline plugins/discovery-gce/src/test/java/org/elasticsearch/cloud/gce/GceInstancesServiceTests.java]: Now we have a QA test do we really need this one?

File: plugins/discovery-gce/src/test/resources/org/elasticsearch/discovery/gce/compute/v1/projects/metadataserver/zones/europe-west1-b/instances [hunk 8]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (tlrx) [inline plugins/discovery-gce/src/test/resources/org/elasticsearch/discovery/gce/compute/v1/projects/metadataserver/zones/europe-west1-b/instances]: I think that we should replace those JSON files by simple code in the test class (not in this PR). I'm not even sure that they are all used.

File: plugins/discovery-gce/qa/gce/build.gradle [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    inputs.properties(expansions)
+    MavenFilteringHack.filter(it, expansions)
+}
+
+integTestCluster {
+    dependsOn gceFixture
+    numNodes = gceNumberOfNodes
+    plugin ':plugins:discovery-gce'
+    setting 'discovery.zen.hosts_provider', 'gce'
+    setting 'cloud.gce.host', "http://${-> gceFixture.addressAndPort}"
```
Comment (tlrx) [inline plugins/discovery-gce/qa/gce/build.gradle]: Can you add a small comment about what is each setting here?

**Review by tlrx — COMMENTED** [review]
It looks good, I left some minor comments. That's a very good thing that the QA test spotted an issue during manual tests, well done!

> One think I don't feel quite comfortable is to exposing two new settings - it is not possible to run QA cluster w/o them

I agree... Maybe we could use a system property here (something similar to es.allow_insecure_settings) that would register the test settings only when the system property is set? In regular test classes you could still extend the GceDiscoveryPlugin to register the extra settings.

@ywelsch Can you give an opinion? And look at the PR too? Thanks :)

File: plugins/discovery-gce/qa/gce/build.gradle (hunk 9 above)
Comment (vladimirdolzhenko) [reply plugins/discovery-gce/qa/gce/build.gradle]: done aa7c93b

File: plugins/discovery-gce/src/test/resources/org/elasticsearch/discovery/gce/compute/v1/projects/metadataserver/zones/europe-west1-b/instances (hunk 8 above)
Comment (vladimirdolzhenko) [reply plugins/discovery-gce/src/test/resources/org/elasticsearch/discovery/gce/compute/v1/projects/metadataserver/zones/europe-west1-b/instances]: they are used in GceDiscoveryTests#testMetadataServerValues

https://github.com/elastic/elasticsearch/pull/33721/files#diff-9a2669df7c9d3ea8a5a5100ee4f78951R274

**Comment by vladimirdolzhenko** [comment]
@tlrx in 1c2a66b I've added a system property `es.allow_reroute_gce_settings` to expose for the test purposes those two GCE-specific settings, so they are not exposed by default

File: plugins/discovery-gce/src/main/java/org/elasticsearch/plugin/discovery/gce/GceDiscoveryPlugin.java [hunk 10]
```diff
@@ -49,8 +51,12 @@
 
 public class GceDiscoveryPlugin extends Plugin implements DiscoveryPlugin, Closeable {
 
+    /** Determines whether settings those reroutes GCE call should be allowed. */
```
Comment (tlrx) [inline plugins/discovery-gce/src/main/java/org/elasticsearch/plugin/discovery/gce/GceDiscoveryPlugin.java]: Can you add that this is for testing purpose only?

**Review by tlrx — APPROVED** [review]
LGTM, thanks @vladimirdolzhenko

File: docs/plugins/discovery-gce.asciidoc [hunk 11]
```diff
@@ -26,12 +26,17 @@ The following gce settings (prefixed with `cloud.gce`) are supported:
...
```
Comment (ywelsch) [inline docs/plugins/discovery-gce.asciidoc]: Simpler: 
> By default the project id will be derived from the instance metadata

File: docs/plugins/discovery-gce.asciidoc [hunk 12]
```diff
@@ -26,12 +26,17 @@ The following gce settings (prefixed with `cloud.gce`) are supported:
...
+
+     Note: it does not utilize any system properties or environment variables like `GOOGLE_CLOUD_PROJECT` or `GCLOUD_PROJECT`
```
Comment (ywelsch) [inline docs/plugins/discovery-gce.asciidoc]: > Note: Deriving the project id from system properties or environment variables (`GOOGLE_CLOUD_PROJECT` or `GCLOUD_PROJECT`) is not supported.

File: docs/plugins/discovery-gce.asciidoc [hunk 13]
```diff
@@ -26,12 +26,17 @@ The following gce settings (prefixed with `cloud.gce`) are supported:
...
+     to resolve the default project id.
 
  `zone`::
 
-     helps to retrieve instances running in a given zone (mandatory). It should be one of the
-     https://developers.google.com/compute/docs/zones#available[GCE supported zones].
+     helps to retrieve instances running in a given zone.
+     It should be one of the https://developers.google.com/compute/docs/zones#available[GCE supported zones].
+     If the value is not specified, default zone is picked up from a metadata server.
```
Comment (ywelsch) [inline docs/plugins/discovery-gce.asciidoc]: Same here:
> By default the zone will be derived from the instance metadata

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java [hunk 14]
```diff
@@ -180,6 +234,16 @@ public synchronized Compute client() {
         return this.client;
     }
 
+    @Override
+    public String project() {
```
Comment (ywelsch) [inline plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: call this method `projectId`?

**Review by ywelsch — APPROVED** [review]

File: plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java (hunk 14 above)
Comment (vladimirdolzhenko) [reply plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceInstancesServiceImpl.java]: addressed in 38cdcc0

File: docs/plugins/discovery-gce.asciidoc (hunk 13 above)
Comment (vladimirdolzhenko) [reply docs/plugins/discovery-gce.asciidoc]: thanks, fixed in 1159638

File: plugins/discovery-gce/src/main/java/org/elasticsearch/plugin/discovery/gce/GceDiscoveryPlugin.java (hunk 10 above)
Comment (vladimirdolzhenko) [reply plugins/discovery-gce/src/main/java/org/elasticsearch/plugin/discovery/gce/GceDiscoveryPlugin.java]: fixed in cb5ae5f

**Comment by vladimirdolzhenko** [comment]
thanks @tlrx and @ywelsch

**Pull request merged**